hpx/runtime/parcelset/parcel.hpp

```diff
@@ -9,698 +10,343 @@
+                    >::value
+                >::type
+          >
+        parcel(
+            naming::id_type const& dests,
+            naming::address const& addrs,
+            Continuation && cont,
+            Action,
+            Args &&... args
+        )
 #if defined(HPX_SUPPORT_MULTIPLE_PARCEL_DESTINATIONS)
-        parcel(std::vector<naming::id_type> const& apply_to,
-                std::vector<naming::address> const& addrs,
-                actions::action_type act)
-          : data_(new detail::multi_destination_parcel_data(apply_to, addrs, act))
-        {
-        }
+          : dests_(1, dests),
+            addrs_(1, addrs),
+#else
+          : dests_(dests),
+            addrs_(addrs),
 #endif
-
-        parcel(naming::id_type const& apply_to,
-                naming::address const& addrs, actions::base_action* act,
-                actions::continuation* do_after)
-          : data_(new detail::single_destination_parcel_data(apply_to, addrs, act, do_after))
+            cont_(
+                new typename util::decay<Continuation>::type(std::forward<Continuation>(cont))
+            ),
+            action_(new actions::transfer_action<Action>(std::forward<Args>(args)...))
+        {
+            HPX_ASSERT(components::types_are_compatible(
+                action_->get_component_type(), this->addrs()[0].type_
+            ));
+        }
+
+        template <
+            typename Action
+          , typename...Args
+          , typename Enable
+                = typename std::enable_if<
+                    traits::is_action<Action>::value
+                >::type
+          >
+        parcel(
+            naming::id_type const& dests,
+            naming::address const& addrs,
+            std::unique_ptr<actions::continuation> cont,
+            Action,
+            Args &&... args
+        )
+#if defined(HPX_SUPPORT_MULTIPLE_PARCEL_DESTINATIONS)
+          : dests_(1, dests),
+            addrs_(1, addrs),
+#else
+          : dests_(dests),
+            addrs_(addrs),
+#endif
+            cont_(std::move(cont)),
+            action_(new actions::transfer_action<Action>(std::forward<Args>(args)...))
         {
+            HPX_ASSERT(components::types_are_compatible(
+                action_->get_component_type(), this->addrs()[0].type_
+            ));
         }
 
-        parcel(naming::id_type const& apply_to,
-                naming::address const& addrs, actions::base_action* act,
-                actions::continuation_type do_after)
-          : data_(new detail::single_destination_parcel_data(apply_to, addrs, act, do_after))
-        {
-        }
+        parcel(parcel && other)
+          : data_(std::move(other.data_)),
+            source_id_(std::move(other.source_id_)),
+            dests_(std::move(other.dests_)),
+            addrs_(std::move(other.addrs_)),
+            cont_(std::move(other.cont_)),
+            action_(std::move(other.action_))
+       {}
```

Comment:
I will check again if everything still works with removing it. While making those changes I needed it at least one point in time. Do we have `parcel(parcel && other) = default` available?